### DISCUSSION TITLE
Callr error when running tar_make() in renv project

### DISCUSSION THREAD

**Discussion by adamaltmejd** [ropensci/targets#506] [Help]
Not sure if this is a bug with targets, callr or renv to be honest, but trying here first. 

I have been running in to this error message for a while now at the end of the make process but just ignoring it since everything seems to be working anyways:

```
Error in dyn.unload(env$.path) :
  shared object '/tmp/RtmpAf1Tsp/callr-client--2e18fe9.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
```

But I also had the setting `TAR_MAKE_REPORTER="timestamp"` set in my .Renviron and when I removed it `tar_make()` refuses to run and produces the following error:

```
Error in if (!enabled) return(FALSE) :
  missing value where TRUE/FALSE needed
Calls: source ... renv_bootstrap_load -> <Anonymous> -> renv_load_rprofile
Traceback (most recent calls last):
17: source(".Rprofile", local = TRUE)
16: withVisible(eval(ei, envir))
15: eval(ei, envir)
14: eval(ei, envir)
13: source("renv/activate.R")
12: withVisible(eval(ei, envir))
11: eval(ei, envir)
10: eval(ei, envir)
 9: local(...)
 8: eval.parent(substitute(eval(quote(expr), envir)))
 7: eval(expr, p)
 6: eval(expr, p)
 5: eval(quote(...), new.env())
 4: eval(quote(...), new.env())
 3: renv_bootstrap_load(project, libpath, version)
 2: renv::load(project)
 1: renv_load_rprofile(project)
Execution halted
Error: callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
```

I have no problem starting the session or manually running `source("renv/activate.R")` so it seems it only relates to the process started by callr. This is confirmed by `tar_make(callr_function = NULL)` running without problem.

It turns out the issue is somehow related to the setting `RENV_CONFIG_USER_PROFILE=TRUE` in my `.Renviron`, but not to the contents of the `~/.Rprofile` since the error happens even when `~/.Rprofile` is empty. But as long as `RENV_CONFIG_USER_PROFILE` is set at all (either to `TRUE` or `FALSE`) the error shows up. But if I have `TAR_MAKE_REPORTER="timestamp"`, `tar_make()` runs but the prints the error at the end. What do you think, is it a targets issue or not?

Edit: just realized I get the same error running `drake::r_make()` in a drake project, so seems less likely its targets related I guess.

**Comment by wlandau** [comment]
It seems like this issue should be related to your `renv` configuration rather than `targets`. To test, you can try activating the `renv` library from inside a `callr` process.

```r
callr::r(function() library(tidyverse))
```

I am betting you will get the similar errors. Although, I cannot explain what `TAR_MAKE_REPORTER` might have to do with it. Incidentally, I recently deprecated `TAR_MAKE_REPORTER` in favor of `tar_config_get("reporter_make")` in the development version.

**Reply by adamaltmejd** [reply]
Actually not, don't have tidyverse installed but

```
callr::r(function() library(targets))
```

runs through and outputs:

```
[1] "targets"   "stats"     "graphics"  "grDevices" "datasets"  "utils"     "methods"   "base"
```

I'll stop using `TAR_MAKE_REPORTER` as soon as it doesn't stop the whole make process :). Will try to report this on the `renv` github. Edit: or maybe `callr` first actually.

**Reply by adamaltmejd** [reply]
Reported here: https://github.com/r-lib/callr/issues/207

**Reply by wlandau** [reply]
I am surprised nothing errored out that time. Do you have a reproducible example you can share?

**Reply by adamaltmejd** [reply]
Hmm.. so best I can do: 

`_targets.R`
```
library(targets)
tar_option_set(
    library = .libPaths(),
    packages = c("tidyr")
)
list(tar_target(x, 1 + 1))
```

with empty ~/.Renviron and ~/.Rprofile this runs well after i set .libPaths() manually to my library. Then i run `renv::init()`, restart R, run `tar_destroy()` and `tar_make()` and it still works. Then I add the following line to `~/.Renviron`:

```
RENV_CONFIG_USER_PROFILE=FALSE
```

And when I run `tar_make()` again:

```
> tar_make()
Error in if (!enabled) return(FALSE) : 
  missing value where TRUE/FALSE needed
Calls: source ... renv_bootstrap_load -> <Anonymous> -> renv_load_rprofile
Traceback (most recent calls last):
17: source(".Rprofile", local = TRUE)
16: withVisible(eval(ei, envir))
15: eval(ei, envir)
14: eval(ei, envir)
13: source("renv/activate.R")
12: withVisible(eval(ei, envir))
11: eval(ei, envir)
10: eval(ei, envir)
 9: local(...)
 8: eval.parent(substitute(eval(quote(expr), envir)))
 7: eval(expr, p)
 6: eval(expr, p)
 5: eval(quote(...), new.env())
 4: eval(quote(...), new.env())
 3: renv_bootstrap_load(project, libpath, version)
 2: renv::load(project)
 1: renv_load_rprofile(project)
Execution halted
Error: callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Visit https://books.ropensci.org/targets/debugging.html for debugging advice.
```

But `tar_make(callr_function=NULL)` works as before.

**Reply by adamaltmejd** [reply]
Btw this is on a Mac 

> sessionInfo()
R version 4.1.0 (2021-05-18)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Big Sur 11.3

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRblas.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.1-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] targets_0.5.0.9000

loaded via a namespace (and not attached):
 [1] igraph_1.2.6      rstudioapi_0.13   knitr_1.33        magrittr_2.0.1   
 [5] tidyselect_1.1.1  R6_2.5.0          rlang_0.4.11      fansi_0.5.0      
 [9] tools_4.1.0       data.table_1.14.1 xfun_0.23         utf8_1.2.1       
[13] cli_2.5.0         withr_2.4.2       ellipsis_0.3.2    yaml_2.2.1       
[17] digest_0.6.27     tibble_3.1.2      lifecycle_1.0.0   crayon_1.4.1     
[21] processx_3.5.2    purrr_0.3.4       callr_3.7.0       vctrs_0.3.8      
[25] ps_1.6.0          codetools_0.2-18  glue_1.4.2        compiler_4.1.0   
[29] pillar_1.6.1      pkgconfig_2.0.3

**Reply by adamaltmejd** [reply]
Reported on renv repo too now: https://github.com/rstudio/renv/issues/773

**Reply by wlandau** [reply]
> with empty ~/.Renviron and ~/.Rprofile 

How exactly are your `.Rprofile` files set up? Normally, `renv` requires a project-level `.Rprofile` with `source("renv/activate.R")`. I am not sure `RENV_CONFIG_USER_PROFILE=false` is actually necessary because `callr` ignores the user-level `~/.Rprofile` by default.

**Reply by wlandau** [reply]
I tried setting up a new project with the `_targets.R` file from https://github.com/ropensci/targets/discussions/506#discussioncomment-847612 and setting RENV_CONFIG_USER_PROFILE in various places but could not reproduce the error.

**Reply by adamaltmejd** [reply]
So the site (project directory) `.Rprofile` had only the line `source("renv/activate.R")`. For creating the example above I removed everything in my `~/.Rprofile`. So I don't think what I regularly keep there should matter.

**Reply by adamaltmejd** [reply]
Very strange. I set `RENV_CONFIG_USER_PROFILE` in my home folder `.Renviron`. Which then in the example above only contained the line `RENV_CONFIG_USER_PROFILE=false`.

**Reply by adamaltmejd** [reply]
That callr ignores user level `~/.Rprofile` is probably a clue though. Maybe `renv` tries to look for it if RENV_CONFIG_USER_PROFILE is set (even when its set to false)  but somehow can't find it from within the callr session.

**Reply by wlandau** [reply]
Yup, seems like a clue. I recommend removing RENV_CONFIG_USER_PROFILE. There does not seem like a reason to keep it, and removing it should make the error go away.

**Reply by adamaltmejd** [reply]
Its messy. Would need to set options like .libPaths() and other custom startup things in the site `.Rprofile` but thats committed in git and used by collaborators. So not really a working solution.

**Reply by wlandau** [reply]
But you also had the same problem when TAR_MAKE_REPORTER was set in your .Renviron, right? Could it be your setup doesn't like ~/.Renviron at all?

**Reply by adamaltmejd** [reply]
Ok some progress! TAR_MAKE_REPORTER actually does not matter at all. The reason I was getting another error (where tar_make() wasn't even running) when TAR_MAKE_REPORTER was set was that renv needs there to be a newline in `~.Renviron` after RENV_CONFIG_USER_PROFILE when loaded from within callr. If there is no newline starting R still loads my `~/.Rprofile` but `tar_make()` fails. As before `callr::r(function() library(targets))` works though. 

If there is a newline and `RENV_CONFIG_USER_PROFILE=FALSE` everything runs like it should. But with a newline and `RENV_CONFIG_USER_PROFILE=TRUE` every run of tar_make() (no matter if it builds anything or not) produces an error at the end of the processs. Everything runs through though.

Btw I'm getting the same error both on Mac and Linux.

```
Error in dyn.unload(env$.path) : 
  shared object '/var/folders/b9/b7hn3x0x7pv_1q51k3xmw6m40000gn/T//RtmpFN9t2P/callr-client--72f7508.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
```

**Reply by adamaltmejd** [reply]
Here is an repo that should produce the errors: https://github.com/adamaltmejd/renv_callr_targets_error

Clone and run run.sh, should produce both errors.

**Reply by wlandau** [reply]
Thanks for the reprex, I can reproduce both errors now. With `Renviron_error1`, I also got an error using just `callr` and not `targets`:

```r
> callr::r(function() 1)
Error in get_result(output = out, options) : 
  callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Type .Last.error.trace to see where the error occurred
> .Last.error
<callr_status_error in get_result(output = out, options):
 callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed>
> .Last.error.trace

 Stack trace:

 1. callr::r(function() 1)
 2. callr:::get_result(output = out, options)
 3. throw(new_callr_error(output, killmsg))

 x callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed 
```

Still working on the other one.

**Reply by wlandau** [reply]
Update: using https://github.com/adamaltmejd/renv_callr_targets_error, I can reproduce both errors with `show = TRUE` in `callr::r()`. So I believe `targets` is not the cause of either error.

## First error

Using your first two dotfiles, particularly the `~/.Renviron` with no line break:

```
mv ~/.Renviron ~/.Renviron_old
cp Renviron_error1 ~/.Renviron

mv ~/.Rprofile ~/.Rprofile_old
cp Rprofile_empty ~/.Rprofile
```

Here is what I see:

```r
> callr::r(function() TRUE, show = TRUE)
Error in if (!enabled) return(FALSE) : 
  missing value where TRUE/FALSE needed
Calls: source ... renv_bootstrap_load -> <Anonymous> -> renv_load_rprofile
Traceback (most recent calls last):
17: source(".Rprofile", local = TRUE)
16: withVisible(eval(ei, envir))
15: eval(ei, envir)
14: eval(ei, envir)
13: source("renv/activate.R")
12: withVisible(eval(ei, envir))
11: eval(ei, envir)
10: eval(ei, envir)
 9: local(...)
 8: eval.parent(substitute(eval(quote(expr), envir)))
 7: eval(expr, p)
 6: eval(expr, p)
 5: eval(quote(...), new.env())
 4: eval(quote(...), new.env())
 3: renv_bootstrap_load(project, libpath, version)
 2: renv::load(project)
 1: renv_load_rprofile(project)
Execution halted
Error in get_result(output = out, options) : 
  callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed
Type .Last.error.trace to see where the error occurred
In addition: Warning message:
could not retrieve available packages for url 'https://Rdatatable.gitlab.io/data.table/bin/macosx/contrib/4.0' 
> .Last.error
<callr_status_error in get_result(output = out, options):
 callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed>
> .Last.error.trace

 Stack trace:

 1. callr::r(function() TRUE, show = TRUE)
 2. callr:::get_result(output = out, options)
 3. throw(new_callr_error(output, killmsg))

 x callr subprocess failed: could not start R, exited with non-zero status, has crashed or was killed 
```

## Second error

Using the `~/.Renviron` file with the line break at the end:

```
cp Renviron_error2 ~/.Renviron
```

Here is what I see:

```r
> callr::r(function() TRUE, show = TRUE)
Error in dyn.unload(env$.path) : 
  shared object '/var/folders/k3/q1f45fsn4_13jbn0742d4zj40000gn/T//Rtmp6xlzNZ/callr-client--4669635.so' was not loaded
Calls: <Anonymous> -> <Anonymous> -> dyn.unload
[1] TRUE
```

**Reply by adamaltmejd** [reply]
Thanks for engaging with this. Seems clear its on either callr or renv.
